api/db.py

```diff
@@ -68,11 +69,13 @@ async def find_by_attributes(self, model, attributes):
         """Find objects with matching attributes
 
         Find all objects with attributes matching the key/value pairs in the
-        attributes dictionary and return them as a list.
+        attributes dictionary using pagination and return the paginated
+        response.
+        The response dictionary will include 'items', 'total', 'limit',
+        and 'offset' keys.
         """
-        col = self._get_collection(model)
-        data = await col.find(attributes).to_list(None)
-        return list(model(**obj) for obj in data)
+        return await paginate(collection=self._get_collection(model),
+                              query_filter=attributes)
```

Comment:
Looks good. It would be interesting to get some stats from Mongo DB about the number of objects looked up or the time it took to complete the query, or to trace the actual operations it's doing.  Overall I guess it should already speed up how the email report is reading all the nodes currently.

Reply:
This is the implementation of `paginate` method:
```
async def paginate(
    collection: AsyncIOMotorCollection,
    query_filter: Optional[Dict[Any, Any]] = None,
    params: Optional[AbstractParams] = None,
    **kwargs: Any,
) -> AbstractPage:
    params = resolve_params(params)
    query_filter = query_filter or {}

    raw_params = params.to_raw_params()
    total = await collection.count_documents(query_filter)
    cursor = collection.find(query_filter, skip=raw_params.offset, limit=raw_params.limit, **kwargs)
    items = await cursor.to_list(length=raw_params.limit)

    return create_page(items, total, params)
```
Basically, it uses a `find` query with `skip` and `limit` parameters at the backend.

Reply:
Below are some statistics:

1. Without pagination:
```
data = await col.find(attributes, skip, limit).to_list(None)
list(model(**obj) for obj in data)
```
- Execution time: 
total nodes = 50                :0.009965658187866211 sec
total nodes=50, offset=2   :0.010911703109741211 sec
total nodes=50, limit=35   :0.013278484344482422 sec

2. Using `paginate`:
```
await paginate(collection=col, query_filter=attributes)
```
- Execution time:
total nodes = 50                :0.004123210906982422 sec
total nodes=50, offset=2   :0.004777431488037109 sec
total nodes=50, limit=35   :0.004402637481689453 sec

Reply:
Nice.  I guess the time would be less when fetching a smaller number of nodes? (either lower limit or last page)

Reply:
Yes, below are the numbers:
**1. Without pagination:**

- total:50 limit:10
  execution time: 0.007180929183959961

- total:50 offset:48
  execution time: 0.008078813552856445	
	
- total:92 offset:50 limit=50
  execution time: 0.009499788284301758

**2.  With pagination:**

- total:50 limit:10
  execution time 0.003947257995605469
	
- total:50 offset:48 limit=50
  execution time: 0.0029668807983398438
	
- total:92 offset:50 limit=50
  execution time: 0.005255937576293945

Reply:
OK I think all these numbers are very small and the difference based on the number of nodes is actually noise compared to the rest of the time to perform a request.  Well I think we'll need some end-to-end testing to really test entire features like that, at least it seems to be working as expected.

Reply:
With "no pagination", I guess that means without the Mongo DB paginated queries?  I see there's still the limit / offset arguments but I guess they were just passed to the API and not the DB query?

Reply:
At the moment, I only have 50 to 100 nodes in the DB, that's why we can not get a massive difference in the performance.

Reply:
Yes, we'll get some more realistic results with the staging instance.